ui/app/components/secret-create-or-update.js

```diff
@@ -0,0 +1,249 @@
+/**
+ * @module SecretCreateOrUpdate
+ * SecretCreateOrUpdate component displays either the form for creating a new secret or creating a new version of the secret
+ *
+ * @example
+ * ```js
+ * <SecretCreateOrUpdate
+ *  @mode="create"
+ *  @model={{model}}
+ *  @showAdvancedMode=true
+ *  @modelForData={{@modelForData}}
+ *  @isV2=true
+ *  @secretData={{@secretData}}
+ *  @canCreateSecretMetadata=true
+ * />
+ * ```
+ * @param {string} mode - create, edit, show determines what view to display
+ * @param {object} model -  the route model, comes from secret-v2 ember record
+ * @param {boolean} showAdvancedMode - whether or not to show the JSON editor
+ * @param {object} modelForData - a class that helps track secret data, defined in secret-edit
+ * @param {boolean} isV2 - whether or not KV 1 or KV 2
+ * @param {object} secretData - class that is created in secret-edit
+ * @param {boolean} canCreateSecretMetadata - based on permissions to the /metadata/ endpoint. If create access.
+ */
+
+import Component from '@glimmer/component';
+import ControlGroupError from 'vault/lib/control-group-error';
+import Ember from 'ember';
+import keys from 'vault/lib/keycodes';
+
+import { action } from '@ember/object';
+import { inject as service } from '@ember/service';
+import { set } from '@ember/object';
+import { tracked } from '@glimmer/tracking';
+
+import { isBlank, isNone } from '@ember/utils';
+import { task, waitForEvent } from 'ember-concurrency';
+
+const LIST_ROUTE = 'vault.cluster.secrets.backend.list';
+const LIST_ROOT_ROUTE = 'vault.cluster.secrets.backend.list-root';
+const SHOW_ROUTE = 'vault.cluster.secrets.backend.show';
+export default class SecretCreateOrUpdate extends Component {
+  @tracked codemirrorString = null;
+  @tracked error = null;
+  @tracked secretPaths = null;
+  @tracked validationErrorCount = 0;
+  @tracked validationMessages = null;
+
+  @service controlGroup;
+  @service router;
+  @service store;
+  @service wizard;
+
+  constructor() {
+    super(...arguments);
+    this.codemirrorString = this.args.secretData.toJSONString();
```

Comment:
Essentially takes the incoming empty object and sets it to a JSON string that is rendered in the JSON view when toggled. Because `secretEdit` is responsible for managing the `secretData` class, I pass this into the component instead of define it here.

Reply:
I also alphabetized the functions as groups, so they were easier to find. I found myself getting lost in the various calls and this was the easiest way for me to organize things.